### PR TITLE
API and code improvements

### PR DESCRIPTION
**Pull request by jsign**
This PR contains multiple changes, such as:
- Document all exported methods in important packages so clients can rely on them.
- Refactor/Renaming of methods and constants with (IMHO) better alternatives.
- Fixed a performance TODO in `MultiExp`.
- Improved CI.
- Improved README, giving particular advice on how clients should use this repository.

I'll be creating a `go-verkle` PR to feel how this looks like from a client-side and get some feedback. So, leaving this as a draft for a bit for now.

Also, I've already updated a bit the repo top (description, web, tags):
image

I'm planning to take care of all `panics(...)` in a further PR.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+- Multiproof prover and verifier implementations.
+
+## Usage in Verkle Tree client libraries
+
+It's extremely important that clients using this library for Verkle Tree implementations only use the following packages:
+- `common` for general utility functions.
+- `banderwagon` for the prime-order group.
+- `ipa` for proof generation and verification.
+
+**Do not** use the `bandersnatch` package directly nor use unsafe functions to get into `banderwagon` internals. Doing so can create a security vulnerability in your implementation.
```
[the remark this answers is not in the document]
Comment (kevaundray) [reply README.md:35]: In the future, we could possibly just have an internal package

File: ipa/config.go [hunk 2]
```diff
@@ -173,8 +177,10 @@ func compute_num_rounds(vector_size uint32) uint32 {
 	return uint32(res)
 }
 
+// GenerateRandomPoints generates numPoints random points on the curve using
+// hardcoded seed.
 func GenerateRandomPoints(numPoints uint64) []banderwagon.Element {
-	seed := "eth_verkle_oct_2021" // incase it changes or needs updating, we can use eth_verkle_month_year
+	seed := "eth_verkle_oct_2021"
```
[the remark this answers is not in the document]
Comment (kevaundray) [reply ipa/config.go:183]: Yeah I think I should have avoided the oct_2021 initially :)

File: README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (jsign) [reply README.md:35]: Yeah, I think it mostly depends on defining if the repo is fully focused in VKTs or wants to also be "open" for people using bandersnatch in general.

File: banderwagon/element.go [hunk 3]
```diff
@@ -131,20 +137,21 @@ func ElementsToBytes(elements []*Element) [][sizePointCompressed]byte {
 	return serialised_points
 }
 
-func ElementsToBytesUncompressed(elements []*Element) [][sizePointUncompressed]byte {
+// BatchToBytesUncompressed serialises a slice of group elements in uncompressed form.
+func BatchToBytesUncompressed(elements ...*Element) [][UncompressedSize]byte {
```
[the remark this answers is not in the document]
Comment (kevaundray) [reply banderwagon/element.go:141]: Looks great, nit: can probably bikeshed this further in a seperate PR to include the types, ie `BatchToBytesUncompressed` requires me to know that uncompressed is only for elements or to look at the func signature

File: banderwagon/element.go [hunk 4]
```diff
@@ -263,84 +277,87 @@ func MultiMapToScalarField(result []*fr.Element, elements []*Element) {
 	}
 }
 
-// TODO: change this to not use pointers
```
[the remark this answers is not in the document]
Comment (kevaundray) [reply banderwagon/element.go:266]: Fair point!

File: banderwagon/element.go [hunk 5]
```diff
@@ -263,84 +277,87 @@ func MultiMapToScalarField(result []*fr.Element, elements []*Element) {
...
 
 	x2 := other.inner.X
 	y2 := other.inner.Y
 
+	if x1.IsZero() && y1.IsZero() {
+		return false
+	}
+	if x2.IsZero() && y2.IsZero() {
+		return false
+	}
```
[the remark this answers is not in the document]
Comment (kevaundray) [reply banderwagon/element.go:293]: Yep, this is only possible if for some reason, the points are not initialized properly and are just zeroed out in memory.

**Review by kevaundray — APPROVED** [review]

**Comment by kevaundray** [comment]
This PR is amazing

**Pull request merged**